README.md

```diff
@@ -104,7 +104,98 @@ module.exports = {
 
 Note that while the above example tests a webpage with [puppeteer](https://github.com/GoogleChrome/puppeteer) and uses pentf's has native support for HTTP requests (in `net_utils`) and email sending (in `email`), tests can be anything – they just have to fail the promise if the test fails.
 
-Have a look in the [API documentation](https://boxine.github.io/pentf/index.html) for various helper functions.
+Have a look in the [API documentation](https://boxine.github.io/pentf/) for various helper functions.
+
+## Tips for writing good tests
+
+By their very nature, end-to-end tests can be flaky, i.e. sometimes succeed and sometimes fail when run multiple times. We want the tests to only relay the flakiness of the systems we test, and not introduce any additional flakiness ourselves. Here are a few tips for that:
+
+### Use `data-testid` attributes in browser tests
+
+Unless otherwise documented, class names and document structures are subject to change. By setting an explicit `data-testid="foo"` attribute in the code, the developer and tester set up a stable contract.
+
+**Avoid**: `await page.waitForSelector('h1 > div.main-container form p button.large-button');`
+
+**Use**: `await waitForTestId(page, 'comment-button');`
+
+### Always wait a bit
+
+If the tests run quick and a remote system is slow, the browser UI may not update immediately. If there is high local system load (which is necessary for the tests to run quickly), then even local updates may not be immediate:
+
+**Avoid**: `page.$('foo')`, `page.evaluate(() => document.querySelector('foo'))`
+
+**Use**: `page.waitForSelector('foo')`
+
+### Make sure you start waiting _early_
+
+Make sure that an action you are waiting on has not already happend. In particular, `await page.waitForNavigation()` calls should probably be replaced by checks for the new page.
+
+**Avoid**:
+```
```

Comment:
Nit: We can add the correct language for the snippet by adding it right after the opening backticks of the block. This will enable GitHub's syntax highlighter to properly highlight this block. In this case it's ` ```js`

Reply:
Same for the other language blocks in this PR 🎉

Reply:
Added everyhwere.